util/tlgen/lib.py

```diff
@@ -19,3 +19,168 @@ def is_pow2(v):
         t = t * 2
 
     return False
+
+
+def simplify_addr(dev, xbar):
+    """Reduce the number of entries smaller by combining them
+
+    If any contiguous regions exist, concatenate them.
+    For instance, 0x1000 ~ 0x1FFF , 0x2000~ 0x2FFF ==> 0x1000 ~ 0x2FFF
+
+    It also checks if there's no device between the gap, then merge
+    the ranges. For instance:
+
+    {0x4000_0000, 0x1_0000}, {0x4008_0000, 0x1_0000} then it combines two
+    entries into {0x4000_0000, 0x9_0000}
+
+    @param addrs List of Dict[Addr] : {'base_addr':,'size_byte':}
+    """
+
+    addrs = dev["addr_range"]
+    # Sort based on the base addr
+    newlist = sorted(addrs, key=lambda k: int(k['base_addr'], 0))
+    # check if overlap or contiguous
+    result = []
+    for e in newlist:
+        if len(result) == 0:
+            result.append(e)
+            continue
+        # if contiguous
+        if int(e["base_addr"], 0) == int(result[-1]["base_addr"], 0) + int(
+                result[-1]["size_byte"], 0):
+            # update previous entry size
+            result[-1]["size_byte"] = "0x{:x}".format(
+                int(result[-1]["size_byte"], 0) + int(e["size_byte"], 0))
+            continue
+
+        if no_device_in_range(xbar, dev["name"], result[-1], e):
+            # Determine if size can be power of 2 value
+            smallest_addr_gte = get_next_base_addr(e["base_addr"], xbar,
+                                                   dev["name"])
+
+            # Choose next value
+            if smallest_addr_gte == -1:
+                next_value = 0x100000000
+            else:
+                next_value = int(smallest_addr_gte["base_addr"], 0)
+
+            calc_size = int(e["base_addr"], 0) + int(e["size_byte"], 0) - int(
+                result[-1]["base_addr"], 0)
+
+            # find power of 2 if possible
+            size_byte = find_pow2_size(result[-1], calc_size, next_value)
+
+            result[-1]["size_byte"] = "0x{:x}".format(size_byte)
+            continue
+
+        # If overlapping (Should it be consider? TlGen will catch it)
+
+        # Normal case
+        result.append(e)
+
+    # return result
+    return result
+
+
+def no_device_in_range(xbar, name, f, t):
+    """Check if other devices doesn't overlap with the from <= x < to
```

Comment:
Maybe should be:

Check if other devices don't overlap with the range 'from <= x < to'